### PR TITLE
adding yaml build

### PR DESCRIPTION
**Pull request by jmarolf**
Allows our signed build to be configured based on per-branch yaml files

### CODE REVIEW

**Comment by davkean** [comment]
Is this enough to only be in master? Won't this mean we need to fork our build definition again for 16.0 vs 15.8/15.9?

**Comment by tmat** [comment]
+1 for adding it to 15.8 and 15.9 branches as well. Having it in all active branches would make it possible to share the new build definition across these branches.

File: .vsts-ci.yml [hunk 1]
```diff
@@ -0,0 +1,107 @@
+resources:
+- repo: self
+  clean: true
+queue:
+  name: VSEng-MicroBuildVS2017
+  condition: succeeded()
```
Comment (tmat) [inline .vsts-ci.yml]: What is this condition doing?

File: .vsts-ci.yml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    filename: 'build\CIBuild.cmd'
+    arguments: '-configuration $(BuildConfiguration)'
+    modifyEnvironment: false
+    
+- task: PublishTestResults@1
+  displayName: Publish Test Results
+  inputs:
+    testRunner: XUnit
+    testResultsFiles: 'artifacts/$(BuildConfiguration)/TestResults/*.xml'
+    testRunTitle: 'Unit Test Results'
```
Comment (tmat) [inline .vsts-ci.yml]: I think this needs `condition: succeededOrFailed()` in order to publish tests if they fail (the build script returns non-zero exit code if tests fail)

**Comment by Pilchie** [comment]
I'm definitely okay with adding this for 15.9.  I'd like to add it for 15.8, but will need some convincing that the build outputs are close enough to identical...

**Comment by davkean** [comment]
@Pilchie @jmarolf To increase confidence in that, I wonder build both build definitions from the same commit with signing turned off and just verify that the shipping binaries are byte-to-byte identical. I'm making the assumption that signing is not deterministic.

File: .vsts-ci.yml (hunk 2 above)
Comment (jmarolf) [reply .vsts-ci.yml]: fixed

**Comment by jmarolf** [comment]
alright I've created a YAML VSTS build definition [here](https://devdiv.visualstudio.com/DevDiv/_build/index?context=allDefinitions&path=%5C&definitionId=9675).  Signing won't work without M2 approval but I will  attempt to build to identical SHAs and compare them as @davkean suggests.  @tmat what is the best way to disable signing?

**Comment by jmarolf** [comment]
Looking at the diffs between these two builds they are identical except for

- Manifest URLs
- Symbol hashes that include the original build path
- vsix extension data that have a randomly generated path on each build (example: `[installdir]\\Common7\\IDE\\Extensions\\hszeiih4.uw3`)

For folder diff see here file://mlangfs1/public/jmarolf/build-diff/Report.html

One file has differences I cannot explain are the differences in Microsoft.VisualStudio.ProjectSystem.Managed.pdb. However I observed these same changes between two builds of the same SHA using the existing signed build (\\CPVSBUILD\drops\Roslyn\Roslyn-Project-System\DotNet-Project-System\20180725.03 and \ \\CPVSBUILD\drops\Roslyn\Roslyn-Project-System\DotNet-Project-System\20180725.04\)

For pdb diff see here: file://mlangfs1/public/jmarolf/build-diff/Microsoft.VisualStudio.ProjectSystem.Managed.pdb-diff.html

Based on this.  The yaml seems to produce a near identical builds to the existing signed build.

**Comment by tmat** [comment]
The difference between the PDBs in \\CPVSBUILD\drops\Roslyn\Roslyn-Project-System\DotNet-Project-System\20180725.03 and \\CPVSBUILD\drops\Roslyn\Roslyn-Project-System\DotNet-Project-System\20180725.04 is different source file content.

**Comment by tmat** [comment]
> vsix extension data that have a randomly generated path on each build (example: [installdir]\\Common7\\IDE\\Extensions\\hszeiih4.uw3)

it would be good to specify this explicitly in the project that generates this, to avoid non-determinism.

**Comment by jmarolf** [comment]
@tmat 
> The difference between the PDBs is different source file content.

Which source files are different? how do I check that?

**Comment by tmat** [comment]
Run `mdv Microsoft.VisualStudio.ProjectSystem.Managed.pdb > 1.txt` on the two versions of the PDB and diff the text files. You'll see differences in source hashes in document table.

https://dotnet.myget.org/feed/metadata-tools/package/nuget/mdv

**Comment by jmarolf** [comment]
@tmat thanks!

It says the following files have different SHAs in the build:

| 22: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/AdditionalFiles.cs' (#18ed)               
| 23: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/AnalyzerReference.cs' (#1e02)             
| 24: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/AppDesigner.cs' (#2015)                   
| 26: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ApplicationDefinition.cs' (#21d1)         
| 27: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/AssemblyReference.cs' (#22a8)             
| 28: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/Compile.cs' (#2746)                       
| 29: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/CompilerCommandLineArgs.cs' (#2bbe)       
| 2a: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/COMReference.cs' (#2d64)                  
| 2b: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ConfigurationGeneral.cs' (#30c5)          
| 2c: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/GeneralConfiguredBrowseObject.cs' (#3fc7) 
| 2d: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/Content.cs' (#48ae)                       
| 2e: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/CopyUpToDateMarker.cs' (#4b3b)            
| 2f: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/DebuggerGeneral.cs' (#4d1d)               
| 30: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/DotNetCliToolReference.cs' (#4f73)        
| 31: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/EmbeddedResource.cs' (#52db)              
| 32: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/Folder.cs' (#53a9)                        
| 33: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/GeneralBrowseObject.cs' (#5525)           
| 34: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/None.cs' (#61ae)                          
| 35: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/NuGetRestore.cs' (#6282)                  
| 36: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/PackageReference.cs' (#6a6e)              
| 37: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/Page.cs' (#6e11)                          
| 38: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectDebugger.cs' (#6e56)               
| 3a: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/UpToDateCheckInput.cs' (#7121)            
| 3b: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/UpToDateCheckOutput.cs' (#7187)           
| 3c: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/UpToDateCheckBuilt.cs' (#71ec)            
| 3d: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/SourceControl.cs' (#7268)                 
| 3e: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/SpecialFolder.cs' (#72c7)                 
| 3f: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/SubProject.cs' (#7323)                    
| 40: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectReference.cs' (#7432)              
| 41: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/VisualBasic.NamespaceImport.cs' (#749f)   
| 42: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ResolvedAssemblyReference.cs' (#750a)     
| 43: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ResolvedCOMReference.cs' (#7571)          
| 44: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ResolvedProjectReference.cs' (#75dc)      
| 45: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/SdkReference.cs' (#7690)                  
| 46: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ResolvedSdkReference.cs' (#76f6)          
| 47: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ResolvedPackageReference.cs' (#7761)      
| 48: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ResolvedAnalyzerReference.cs' (#77e9)     
| 49: '/_/artifacts/Release/obj/Microsoft.VisualStudio.ProjectSystem.Managed/ResolvedCompilationReference.cs' (#7858)  

Looks like files in the obj directory have different contents

**Comment by davkean** [comment]
It's hard to tell without seeing the diffs, but I believe those are not deterministic, they use a GUID or something, right?

**Comment by jmarolf** [comment]
@davkean correct they are all auto-generated and the output is not deterministic.  I would say I have confidence that this is doing everything the same as our existing build (except for signing which I haven't tested yet)

**Comment by davkean** [comment]
@Pilchie Is that good enough confidence for you?

**Comment by Pilchie** [comment]
Yes.  Thanks everyone for the validation!

Perhaps we can make the switch over the weekend so that we can ensure that we don't lose the ability to get signed builds done?

**Comment by jmarolf** [comment]
Sounds good.  I'll merge this but not set up the signed build till later
